Crash in local storage during shutdown

During node shutdown the following situation can happen because the database handle is set to `nil` in `Close()`:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x410 pc=0xc36ba4]

goroutine 69746310 [running]:
github.com/dgraph-io/badger/v4.(*DB).IsClosed(...)
        github.com/dgraph-io/badger/v4@v4.5.1/db.go:541
github.com/dgraph-io/badger/v4.(*DB).View(0xc00ffe9dd0?, 0x4d58b1?)
        github.com/dgraph-io/badger/v4@v4.5.1/txn.go:800 +0x24
github.com/oasisprotocol/oasis-core/go/runtime/localstorage.(*localStorage).Get(0xc006d3eac8, {0xc01a14aaf0, 0x69, 0x41d845?})
        github.com/oasisprotocol/oasis-core/go/runtime/localstorage/localstorage.go:50 +0xa9
github.com/oasisprotocol/oasis-core/go/runtime/registry.(*runtimeHostHandler).handleHostLocalStorageGet(0xc048ae68c0?, 0xc04573ff08)
        github.com/oasisprotocol/oasis-core/go/runtime/registry/handler.go:240 +0x4a
github.com/oasisprotocol/oasis-core/go/runtime/registry.(*runtimeHostHandler).Handle(0xc000744000, {0x25c3930, 0xc0558fabe0}, 0xc0196d6918)
        github.com/oasisprotocol/oasis-core/go/runtime/registry/handler.go:87 +0x34f
github.com/oasisprotocol/oasis-core/go/runtime/host/protocol.(*connection).handleMessage(0xc01a16d340, {0x25c3930, 0xc0558fabe0}, 0xc0196d6908)
        github.com/oasisprotocol/oasis-core/go/runtime/host/protocol/connection.go:407 +0x382
github.com/oasisprotocol/oasis-core/go/runtime/host/protocol.(*connection).workerIncoming.func2()
        github.com/oasisprotocol/oasis-core/go/runtime/host/protocol/connection.go:475 +0x85
created by github.com/oasisprotocol/oasis-core/go/runtime/host/protocol.(*connection).workerIncoming in goroutine 69746306
        github.com/oasisprotocol/oasis-core/go/runtime/host/protocol/connection.go:468 +0xe7
```

--- a/.changelog/6231.bugfix.md
+++ b/.changelog/6231.bugfix.md
@@ -1,0 +1,1 @@
+go/runtime/localstorage: Fix nil dereference after close

--- a/go/runtime/localstorage/localstorage.go
+++ b/go/runtime/localstorage/localstorage.go
@@ -7,6 +7,7 @@ import (
 	"errors"
 	"fmt"
 	"path/filepath"
+	"sync"
 
 	"github.com/dgraph-io/badger/v4"
 	"github.com/dgraph-io/badger/v4/options"
@@ -39,6 +40,8 @@ type localStorage struct {
 
 	db *badger.DB
 	gc *cmnBadger.GCWorker
+
+	closeOnce sync.Once
 }
 
 func (s *localStorage) Get(key []byte) ([]byte, error) {
@@ -92,13 +95,15 @@ func (s *localStorage) Set(key, value []byte) error {
 }
 
 func (s *localStorage) Stop() {
-	s.gc.Stop()
-	if err := s.db.Close(); err != nil {
-		s.logger.Error("failed to close local storage",
-			"err", err,
-		)
-	}
-	s.db = nil
+	s.closeOnce.Do(func() {
+		s.gc.Stop()
+
+		if err := s.db.Close(); err != nil {
+			s.logger.Error("failed to close local storage",
+				"err", err,
+			)
+		}
+	})
 }
 
 // New creates new untrusted local storage.

